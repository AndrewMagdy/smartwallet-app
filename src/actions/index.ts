import * as accountActions from './account'

export const accountActions