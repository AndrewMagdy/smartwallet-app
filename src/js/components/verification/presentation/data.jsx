import React from 'react'
import Radium from 'radium'

import Cake from 'material-ui/svg-icons/social/cake'
import Person from 'material-ui/svg-icons/social/person'
import Location from 'material-ui/svg-icons/maps/place'
import {List, SelectField, MenuItem} from 'material-ui'

import {
  EditAppBar,
  EditHeader,
  EditListItem,
  SelectCountryItem,
  SelectListItem,
  IconNumber,
  DateListItem
} from './ui'

import {Content, Header} from '../../structure'

@Radium
export default class VerificationDataPresentation extends React.Component {
  static propTypes = {
    verify: React.PropTypes.func,
    showVerifierLocations: React.PropTypes.func,
    change: React.PropTypes.func,
    selectCountry: React.PropTypes.func,
    cancel: React.PropTypes.func,
    showVerifiers: React.PropTypes.func,
    loaded: React.PropTypes.bool,
    focusedGroup: React.PropTypes.string,
    focusedField: React.PropTypes.string,
    setFocused: React.PropTypes.func,
    verifierLocations: React.PropTypes.array,
    showErrors: React.PropTypes.bool,
    showAddress: React.PropTypes.bool,
    physicalAddress: React.PropTypes.array,
    idCard: React.PropTypes.array
  }

  renderField(field) {
    switch (field.key) {
      case 'birthCountry':
      case 'country':
        return this.renderCountryField(field)
      case 'gender':
        return this.renderGenderField(field)
      case 'zip':
      case 'birthPlace':
        return null // handled with birthDate or streetWithNumber
      case 'birthDate':
        return this.renderBirthDate(field)
      case 'streetWithNumber':
        return this.renderStreetAndZipFields(field)
      case 'expirationDate':
        return this.renderDateField(field)
      default:
        return this.renderTextField(field)
    }
  }

  renderGenderField({value, label, valid, key, options, index, icon, group}) {
    return <SelectListItem
      id={key}
      key={key}
      value={value}
      label={label}
      focused={false}
      onFocusChange={() => this.props.setFocused(key, group)}
      types={options}
      onDelete={() => this.props.change(key, '')}
      fullWidth
      enableDelete={value.length > 0}
      onChange={(e, i, v) => this.props.change(key, v)} />
  }

  renderCountryField({value, label, valid, key, options, index, icon, group}) {
    return <SelectCountryItem
      id={key}
      key={key}
      icon={icon}
      value={value}
      label={label}
      onFocusChange={() => {
        this.props.setFocused(key, group)
        this.props.selectCountry(key)
      }}
      types={options}
      onDelete={() => this.props.change(key, '')}
      fullWidth
      focused={false}
      enableDelete={value.length > 0}
      onChange={() => this.props.change(key, '')} />
  }

  renderTextField({value, label, valid, key, index, icon, group}) {
    return <EditListItem
      icon={icon}
      key={key}
      id={key}
      label={label}
      underlineHide={!!value}
      name={key}
      value={value}
      onFocusChange={(field) => this.props.setFocused(field, group)}
      showErrors={!valid && this.props.showErrors}
      valid={valid}
      onChange={(e) => this.props.change(key, e.target.value)}
      focused={this.props.focusedGroup === group && !!icon}
      onDelete={() => this.props.change(key, '')}
      enableDelete={value.length > 0} />
  }

  renderStreetAndZipFields({value, label, valid, key, index, icon, group}) {
    const zip = this.props.physicalAddress[1]
    return (<table key={key}><tbody> <tr>
      <td style={{width: '70%'}} key="0" >
        <div>
          <EditListItem
            id={key}
            icon={icon}
            label={label}
            value={value}
            underlineHide={!!value}
            onFocusChange={(field) => this.props.setFocused(field, group)}
            onChange={(e) => this.props.change(key, e.target.value)}
            focused={this.props.focusedGroup === group} />
        </div>
      </td>
      <td style={{width: '30%'}} key="1">
        <div>
          <EditListItem
            id={zip.key}
            label={zip.label}
            underlineHide={!!zip.value}
            value={zip.value}
            onFocusChange={field => this.props.setFocused(field, zip.group)}
            onChange={(e) => this.props.change(zip.key, e.target.value)}
            onDelete={() => {
              this.props.change(zip.key, '')
              this.props.change(key, '')
            }}
            enableDelete={!!zip.value || !!value} />
        </div>
      </td>
    </tr> </tbody> </table>)
  }

  renderBirthDate({value, label, valid, key, index, icon, group}) {
    const birthPlace = this.props.idCard[index + 1]

    return (<table key={key} style={{width: '100%'}}> <tbody><tr>
      <td key="0" style={{width: '70%'}}>
        <div>
          <DateListItem
            icon={icon}
            label={label}
            value={value || null}
            onFocusChange={(field) => this.props.setFocused(field, group)}
            focused={this.props.focusedGroup === group}
            onChange={(e, date) => this.props.change(key, date)} />
        </div>
      </td>
      <td key="1" style={{width: '30%', position: 'fix'}}>
        <div>
          <EditListItem
            id={birthPlace.key}
            label={birthPlace.label}
            value={birthPlace.value}
            underlineHide={!!value}
            onFocusChange={
              (field) => this.props.setFocused(field, birthPlace.group)}
            onChange={(e) => this.props.change(birthPlace.key, e.target.value)}
            onDelete={() => {
              this.props.change(birthPlace.key, '')
              this.props.change(key, '')
            }}
            enableDelete={!!birthPlace.value || !!value} />
        </div>
      </td>
    </tr></tbody> </table>)
  }

  renderDateField({value, label, valid, key, index, icon, group}) {
    return <DateListItem
      key={key}
      icon={icon}
      label={label}
      name={key}
      value={value}
      onFocusChange={(field) => this.props.setFocused(field, group)}
      showErrors={!valid && this.props.showErrors}
      valid={valid}
      focused={this.props.focusedGroup === group}
      onDelete={() => { this.props.change(key, '') }}
      enableDelete={value.toString().length > 0}
      onChange={(event, date) => this.props.change(key, date)} />
  }

  renderOptionsField({value, label, valid, key, options, group, index, icon}) {
    return (<SelectField
      floatingLabelText={label}
      id={key}
      key={key}
      value={value}
      fullWidth
      focused={this.props.focusedGroup === group}
      onFocusChange={(field) => this.props.setFocused(field, group)}
      onChange={(e, i, v) => this.props.change(key, v)}
      onDelete={() => this.props.change(key, '')}
      autoWidth >
      {options.map((gender, index) => <MenuItem
        key={index}
        value={gender}
        primaryText={gender} />)}
    </SelectField>)
  }

  createIcons() {
    const idCardGroups = this.props.idCard.map(({group}) => group)
    let icons = [IconNumber]
    icons[idCardGroups.indexOf('person')] = Person
    icons[idCardGroups.indexOf('cake')] = Cake
    icons[idCardGroups.length] = Location
    return icons
  }

  render() {
    const icons = this.createIcons()
    const {idCard, physicalAddress, showAddress, verify, cancel} = this.props

    let addressFields = physicalAddress[0]
    if (showAddress) { addressFields = physicalAddress }

    const fields = idCard.concat(addressFields).map(
      (field, index) => this.renderField({...field, index, icon: icons[index]}))

    return (<div>
      <EditAppBar
        title="DATA CHECK"
        rightTitle="VERIFFY"
        onSave={verify}
        onClose={cancel} />
      <Header>STEP 2</Header>
      <Content>
        Please fill in the following ID Card details for further verfication.
      </Content>
      <Content>
        <EditHeader title="ID Card" />
        <List>
          {fields}
        </List>
      </Content>
    </div>)
  }
}
