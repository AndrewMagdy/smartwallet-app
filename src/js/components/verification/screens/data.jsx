import React from 'react'
import {connect} from 'redux/utils'
import Presentation from '../presentation/data'

@connect({
  props: ['verification.data'],
  actions: [
    'simple-dialog:showSimpleDialog',
    'simple-dialog:configSimpleDialog',
    'verification/country:initiateCountrySelectScreen',
    'verification/data:changeIdCardField',
    'verification/data:changePhysicalAddressField',
    'verification/data:setFocusedField',
    'verification/data:setShowAddress',
    'verification/data:verifyData',
    'verification/data:cancel'
  ]
})
export default class VerificationDataScreen extends React.Component {
  static propTypes = {
    data: React.PropTypes.object.isRequired,
    changeIdCardField: React.PropTypes.func.isRequired,
    initiateCountrySelectScreen: React.PropTypes.func.isRequired,
    changePhysicalAddressField: React.PropTypes.func.isRequired,
    setFocusedField: React.PropTypes.func.isRequired,
    setShowAddress: React.PropTypes.func.isRequired,
    verifyData: React.PropTypes.func.isRequired,
    cancel: React.PropTypes.func.isRequired
  }

  render() {
    const {cancel, verifyData, initiateCountrySelectScreen} = this.props
    const {focusedField, focusedGroup, verifierLocations
    } = this.props.data

    return <Presentation
      loaded
      focusedGroup={focusedGroup}
      focusedField={focusedField}
      verify={verifyData}
      verifierLocations={verifierLocations}
      setFocused={(...args) => { this.setFocusedElements(...args) }}
      change={(...args) => { this.change(...args) }}
      cancel={cancel}
      selectCountry={initiateCountrySelectScreen}
      showAddress={this.props.data.idCard.showAddress}
      physicalAddress={this.parseAddressDetailsToArray()}
      idCard={this.parseIdCardDetailsToArray()} />
  }

  change(field, value) {
    const idCardFields = this.parseIdCardDetailsToArray()
      .map(({key}) => key)
    if (idCardFields.includes(field)) {
      return this.props.changeIdCardField(field, value)
    } else if (field === 'streetWithNumber') {
      this.props.setShowAddress(value.trim().length > 0)
    }
    return this.props.changePhysicalAddressField(field, value)
  }

  setFocusedElements(key, group) {
    if (key === '') {
      return this.props.setFocusedField('', '')
    }
    return this.props.setFocusedField(key, group)
  }

  parseIdCardDetailsToArray() {
    const {number, expirationDate, firstName, lastName, gender, birthDate,
      birthPlace, birthCountry} = this.props.data.idCard
    return [
      {label: 'Id Card Number', key: 'number', group: 'numbers', ...number},
      {label: 'Expiration Date', key: 'expirationDate', group: 'numbers', ...expirationDate}, // eslint-disable-line max-len
      {label: 'First Name', key: 'firstName', group: 'person', ...firstName},
      {label: 'Last Name', key: 'lastName', group: 'person', ...lastName},
      {label: 'Gender', key: 'gender', group: 'person', ...gender},
      {label: 'Date of Birth', key: 'birthDate', group: 'cake', ...birthDate},
      {label: 'Place of Birth', key: 'birthPlace', group: 'cake', ...birthPlace}, // eslint-disable-line max-len
      {label: 'Country of Birth', key: 'birthCountry', group: 'cake', ...birthCountry} // eslint-disable-line max-len
    ]
  }

  parseAddressDetailsToArray() {
    const {streetWithNumber, zip, city, state, country} = this.props
      .data.idCard.physicalAddress
    const group = 'address'
    return [
      {...streetWithNumber, key: 'streetWithNumber', label: 'Street', group},
      {...city, key: 'city', label: 'City', group},
      {...zip, key: 'zip', label: 'Zip Code', group},
      {...state, key: 'state', label: 'State', group},
      {...country, key: 'country', label: 'Country', group}
    ]
  }
}
