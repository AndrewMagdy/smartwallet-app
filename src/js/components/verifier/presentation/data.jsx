import React from 'react'
import Radium from 'radium'

import Cake from 'material-ui/svg-icons/social/cake'
import Person from 'material-ui/svg-icons/social/person'
import Location from 'material-ui/svg-icons/maps/place'
import List from 'material-ui/List'

import {
  EditAppBar,
  EditHeader,
  EditListItem,
  SelectCountryItem,
  SelectListItem,
  IconNumber,
  DateListItem
} from './ui'

import {Content, Header} from '../../structure'

const STYLES = {
  container: {
    width: '100%'
  },
  firstField: {
    width: '70%',
    position: 'fix'
  },
  secondField: {
    width: '30%',
    position: 'fix'
  },
  textAlign: {
    textAlign: 'center'
  }
}

@Radium
export default class VerificationDataPresentation extends React.Component {
  static propTypes = {
    verify: React.PropTypes.func,
    showVerifierLocations: React.PropTypes.func,
    change: React.PropTypes.func,
    selectCountry: React.PropTypes.func,
    cancel: React.PropTypes.func,
    showVerifiers: React.PropTypes.func,
    focusedGroup: React.PropTypes.string,
    focusedField: React.PropTypes.string,
    setFocused: React.PropTypes.func,
    verifierLocations: React.PropTypes.array,
    showErrors: React.PropTypes.bool,
    showAddress: React.PropTypes.bool,
    username: React.PropTypes.object,
    physicalAddress: React.PropTypes.array,
    idCard: React.PropTypes.array
  }

  renderField(field) {
    switch (field.key) {
      case 'birthCountry':
      case 'country':
        return this.renderCountryField(field)
      case 'gender':
        return this.renderGenderField(field)
      case 'birthPlace':
      case 'zip':
        return null // handled with birthDate or city
      case 'birthDate':
      case 'city':
        return this.renderTwoFields(field)
      default:
        return this.renderTextField(field)
    }
  }

  renderGenderField({value, label, valid, key, options, index, icon, group}) {
    return <SelectListItem
      id={key}
      key={key}
      value={value}
      label={label}
      focused={false}
      onFocusChange={() => this.props.setFocused(key, group)}
      types={options}
      onDelete={() => this.props.change(key, '')}
      fullWidth
      enableDelete={value.length > 0}
      onChange={(e, i, v) => this.props.change(key, v)} />
  }

  renderCountryField({value, label, valid, key, options, index, icon, group}) {
    return <SelectCountryItem
      id={key}
      key={key}
      icon={icon}
      value={value}
      label={label}
      onFocusChange={() => {
        this.props.setFocused(key, group)
        this.props.selectCountry(key)
      }}
      types={options}
      onDelete={() => this.props.change(key, '')}
      fullWidth
      focused={false}
      enableDelete={value.length > 0}
      onChange={() => this.props.change(key, '')} />
  }

  renderTextField({value, label, valid, key, index, icon, group}) {
    return <EditListItem
      icon={icon}
      key={key}
      id={key}
      label={label}
      underlineHide={!!value}
      name={key}
      value={value}
      onFocusChange={(field) => this.props.setFocused(field, group)}
      showErrors={!valid && this.props.showErrors}
      valid={valid}
      onChange={(e) => this.props.change(key, e.target.value)}
      focused={this.props.focusedGroup === group && !!icon}
      onDelete={() => this.props.change(key, '')}
      enableDelete={value.length > 0} />
  }

  renderTwoFields({value, label, valid, key, index, icon, group}) {
    const secondField = this.props.physicalAddress.map(({key}) => key).includes(key) // eslint-disable-line max-len
      ? this.props.physicalAddress[index - this.props.idCard.length + 1]
      : this.props.idCard[index + 1]
    return (<table key={key} style={STYLES.container} ><tbody><tr>
      <td style={STYLES.firstField} key="0" >
        {this.renderTextField({value, label, valid, key, index, icon, group})}
      </td>
      <td style={STYLES.secondField} key="1">
        {this.renderTextField(secondField)}
      </td>
    </tr></tbody></table>)
  }

  renderDateField({value, label, valid, key, index, icon, group}) {
    return <DateListItem
      key={key}
      icon={icon}
      label={label}
      name={key}
      value={value}
      onFocusChange={(field) => this.props.setFocused(field, group)}
      showErrors={!valid && this.props.showErrors}
      valid={valid}
      focused={this.props.focusedGroup === group}
      onDelete={() => { this.props.change(key, '') }}
      enableDelete={value.toString().length > 0}
      onChange={(event, date) => this.props.change(key, date)} />
  }

  createIcons() {
    const idCardGroups = this.props.idCard.map(({group}) => group)
    let icons = [IconNumber]
    icons[idCardGroups.indexOf('person')] = Person
    icons[idCardGroups.indexOf('cake')] = Cake
    icons[idCardGroups.length] = Location
    return icons
  }

  render() {
    const icons = this.createIcons()
    const {idCard, physicalAddress, showAddress, verify, cancel, username} = this.props // eslint-disable-line max-len

    const addressFields = showAddress ? physicalAddress : physicalAddress[0]

    const fields = idCard.concat(addressFields).map(
      (field, index) => this.renderField({...field, index, icon: icons[index]}))
    const usernameField = this.renderTextField({
      ...username, icon: Person, index: -1})
    return (<div>
      <EditAppBar
        title="DATA CHECK"
        rightTitle="VERIFFY"
        onSave={verify}
        onClose={cancel} />
      <Header style={STYLES.textAlign}>STEP 2</Header>
      <Content style={STYLES.textAlign}>
        Please fill in the following ID Card details for further verfication.
      </Content>
      <Content>
        <EditHeader title="User Name" />
          {usernameField}
        <EditHeader title="ID Card" />
        <List>
          {fields}
        </List>
      </Content>
    </div>)
  }
}
