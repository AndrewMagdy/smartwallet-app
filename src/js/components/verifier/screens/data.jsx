import React from 'react'
import {connect} from 'redux_state/utils'
import Presentation from '../presentation/data'

@connect({
  props: ['verifier.data'],
  actions: [
    'simple-dialog:showSimpleDialog',
    'simple-dialog:configSimpleDialog',
    'verifier/country:initiateCountrySelectScreen',
    'verifier/data:changeIdCardField',
    'verifier/data:changePhysicalAddressField',
    'verifier/data:setFocusedField',
    'verifier/data:setShowAddress',
    'verifier/data:verifyData',
    'verifier/data:setUsername',
    'verifier/data:cancel'
  ]
})
export default class VerificationDataScreen extends React.Component {
  static propTypes = {
    data: React.PropTypes.object.isRequired,
    changeIdCardField: React.PropTypes.func.isRequired,
    initiateCountrySelectScreen: React.PropTypes.func.isRequired,
    changePhysicalAddressField: React.PropTypes.func.isRequired,
    setFocusedField: React.PropTypes.func.isRequired,
    setShowAddress: React.PropTypes.func.isRequired,
    verifyData: React.PropTypes.func.isRequired,
    setUsername: React.PropTypes.func.isRequired,
    cancel: React.PropTypes.func.isRequired
  }

  render() {
    const {cancel, verifyData, initiateCountrySelectScreen} = this.props
    const {focusedField, focusedGroup, verifierLocations, username
    } = this.props.data

    return <Presentation
      loaded
      focusedGroup={focusedGroup}
      focusedField={focusedField}
      verify={verifyData}
      username={{label: 'User Name', key: 'username', group: 'username', value: username}} // eslint-disable-line max-len
      verifierLocations={verifierLocations}
      setFocused={(...args) => { this.setFocusedElements(...args) }}
      change={(...args) => { this.change(...args) }}
      cancel={cancel}
      selectCountry={initiateCountrySelectScreen}
      showAddress={this.props.data.idCard.showAddress}
      physicalAddress={this.parseAddressDetailsToArray()}
      idCard={this.parseIdCardDetailsToArray()} />
  }

  change(field, value) {
    if (field === 'username') {
      return this.props.setUsername(value)
    }
    const idCardFields = this.parseIdCardDetailsToArray().map(({key}) => key)
    if (idCardFields.includes(field)) {
      return this.props.changeIdCardField(field, value)
    } else if (field === 'streetWithNumber') {
      this.props.setShowAddress(value.trim().length > 0)
    }
    return this.props.changePhysicalAddressField(field, value)
  }

  setFocusedElements(key, group) {
    if (key === '') {
      return this.props.setFocusedField('', '')
    }
    return this.props.setFocusedField(key, group)
  }

  parseIdCardDetailsToArray() {
    const {number, expirationDate, firstName, lastName, gender, birthDate,
      birthPlace, birthCountry} = this.props.data.idCard
    return [
      {label: 'Id Card Number', key: 'number', group: 'numbers', ...number},
      {label: 'Expiration Date', key: 'expirationDate', group: 'numbers', ...expirationDate}, // eslint-disable-line max-len
      {label: 'First Name', key: 'firstName', group: 'person', ...firstName},
      {label: 'Last Name', key: 'lastName', group: 'person', ...lastName},
      {label: 'Gender', key: 'gender', group: 'person', ...gender},
      {label: 'Date of Birth', key: 'birthDate', group: 'cake', ...birthDate},
      {label: 'Place of Birth', key: 'birthPlace', group: 'cake', ...birthPlace}, // eslint-disable-line max-len
      {label: 'Country of Birth', key: 'birthCountry', group: 'cake', ...birthCountry} // eslint-disable-line max-len
    ]
  }

  parseAddressDetailsToArray() {
    const {streetWithNumber, zip, city, state, country} = this.props
      .data.idCard.physicalAddress
    const group = 'address'
    return [
      {...streetWithNumber, key: 'streetWithNumber', label: 'Street', group},
      {...city, key: 'city', label: 'City', group},
      {...zip, key: 'zip', label: 'Zip Code', group},
      {...state, key: 'state', label: 'State', group},
      {...country, key: 'country', label: 'Country', group}
    ]
  }
}
