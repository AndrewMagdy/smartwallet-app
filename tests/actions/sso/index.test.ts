import { ssoActions } from 'src/actions/'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { JolocomLib } from 'jolocom-lib'
import ssoData from './data/mockSSOData'
import data from '../registration/data/mockRegistrationData'

describe.only('SSO action creators', () => {
  jest.setTimeout(30000);
  const initialState = {
    account: {
      did: {
        toJS: () => 'mock:did:test '
      }
    },
    sso: {
      activeCredentialRequest: {
        requester: 'did:mock:requester',
        callbackURL: 'http://test.test',
        availableCredentials: []
      }
    }
  }
  const mockStore = configureStore([thunk])(initialState)
  
  beforeEach(() => {
    mockStore.clearActions()
  })

  it('should correctly parse an incoming encoded JWT', async () => {

    const { encodedCredentialRequestPayload, erroneousJWT } = ssoData
    const actionParse = ssoActions.parseJWT(encodedCredentialRequestPayload)
    await actionParse(mockStore.dispatch, mockStore.getState)
    expect(mockStore.getActions()).toEqual(
      [
        { type: 'SET_LOADING', loading: true }
      ])

    mockStore.clearActions()

    const actionError = ssoActions.parseJWT(erroneousJWT)
    await actionError(mockStore.dispatch, mockStore.getState)
    expect(mockStore.getActions()).toEqual(
      [
        { type: 'SET_LOADING', loading: true },
        { type: 'SET_LOADING', loading: false },
        {
          type: 'Navigation/NAVIGATE',
          routeName: 'Exception',
          params: {
            flag: 'default'
          }
        }
      ])
    // try catch statements for different payload types?
  })

  it('should correctly consume CredentialOfferRequests', async () => {

    //issue in this test with whole functionality in try/catch statements

    const { encodedCredentialOfferPayload } = ssoData

    const backendMiddleware = {
      identityWallet: {
        create: jest.fn()
          .mockResolvedValue([
            {
              header: {typ: "JWT", alg: "ES256K"},
              payload: {
                credentialOffer: {
                  challenge: "6345l",
                  callbackURL: "https://demo-sso.jolocom.com/receive/",
                  instant: true,
                  requestedInput: {}
                },
                iat: 1540457039169,
                iss: "did:jolo:5449f86ba2d4016169350a9766363d63a5326c0d2e8b922c2a75a803dd35d801#keys-1",
                typ: "credentialOfferResponse"
              }
            }
          ])
        }
      }

    // const returnedDecodedJwt = await JolocomLib.parse.interactionJSONWebToken.decode(encodedCredentialOfferPayload)
    // const actionConsumeCOR = ssoActions.consumeCredentialOfferRequest(returnedDecodedJwt)
    // await actionConsumeCOR(mockStore.dispatch, mockStore.getState, backendMiddleware)
    // expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('should correctly consume Credential Requests', async() => {
    const { testSignedCredentialDefault } = data
    const { encodedCredentialRequestPayload } = ssoData

    const backendMiddleware = {
      storageLib: {
        get: {
          attributesByType: jest.fn().mockResolvedValue({
            results: [{
              verification: "claimId:6f3008b6",
              values: ["test", "test"],
              fieldName: "givenName"
            }],
            type: ["Credential", "ProofOfNameCredential"]
          }),
          verifiableCredential: jest.fn()
            .mockResolvedValue([JolocomLib.parse.signedCredential.fromJSON(testSignedCredentialDefault)])
        }
      }
    }

    const returnedDecodedJwt = await JolocomLib.parse.interactionJSONWebToken.decode(encodedCredentialRequestPayload)
    const actionConsumeCR = ssoActions.consumeCredentialRequest(returnedDecodedJwt)
    await actionConsumeCR(mockStore.dispatch, mockStore.getState, backendMiddleware)
    expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('should be able to receive external credentials', async () => {
    const { encodedExternalCredentialPayload } = ssoData
    const returnedDecodedJwt = await JolocomLib.parse.interactionJSONWebToken.decode(encodedExternalCredentialPayload)
    const actionConsumeCR = ssoActions.receiveExternalCredential(returnedDecodedJwt)
    await actionConsumeCR(mockStore.dispatch, mockStore.getState)
    expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('should correctly send a CredentialResponse', async () => {

    const { testSignedCredentialDefault } = data
    const { sampleStateVerificationSummary, sampleEncryptedWif, sampleDecryptedWif } = ssoData
    global.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        ok: true,
        json: () => data
      })
    );
    const backendMiddleware = {
      keyChainLib: {
        getPassword: jest.fn().mockResolvedValue('testpassword'),
      },
      encryptionLib: {
        decryptWithPass: jest.fn().mockReturnValue(sampleDecryptedWif)
      },
      storageLib: {
        get: {
          persona: jest.fn().mockResolvedValue([
            {
              did: 'did:jolo:first',
              controllingKey: sampleEncryptedWif
            }
          ]),
          verifiableCredential: jest.fn()
          .mockResolvedValue([JolocomLib.parse.signedCredential.fromJSON(testSignedCredentialDefault)])
        },
      },
      ethereumLib: {
        wifToEthereumKey: jest.fn()
        .mockReturnValue({
          privateKey:'a9b4cbc2e0cd25e8324ec8018e62219704531173ebbc7ad517bba8d3a9186142',
          address: "test"
        })
      }
    }
    const actionSendCR = ssoActions.sendCredentialResponse(sampleStateVerificationSummary)
    await actionSendCR(mockStore.dispatch, mockStore.getState, backendMiddleware)
    expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('should correctly cancel submission of claims', async () => {
    const action = ssoActions.cancelSSO()
    await action(mockStore.dispatch, mockStore.getState)
    expect(mockStore.getActions()).toMatchSnapshot()
  })

  it('should correctly cancel/go back from an SSO action ', async () => {
    const action = ssoActions.cancelReceiving()
    await action(mockStore.dispatch, mockStore.getState)
    expect(mockStore.getActions()).toMatchSnapshot()
  })
})
