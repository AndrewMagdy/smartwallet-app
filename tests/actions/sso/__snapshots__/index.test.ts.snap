// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSO action creators should be able to receive external credentials 1`] = `
Array [
  Object {
    "external": Array [
      Object {
        "@context": Array [
          Object {
            "Credential": "cred:Credential",
            "EcdsaKoblitzSignature2016": "sec:EcdsaKoblitzSignature2016",
            "claim": Object {
              "@id": "cred:claim",
              "@type": "@id",
            },
            "created": Object {
              "@id": "dc:created",
              "@type": "xsd:dateTime",
            },
            "creator": Object {
              "@id": "dc:creator",
              "@type": "@id",
            },
            "cred": "https://w3id.org/credentials#",
            "credential": Object {
              "@id": "cred:credential",
              "@type": "@id",
            },
            "dc": "http://purl.org/dc/terms/",
            "domain": "sec:domain",
            "expires": Object {
              "@id": "sec:expiration",
              "@type": "xsd:dateTime",
            },
            "id": "@id",
            "issued": Object {
              "@id": "cred:issued",
              "@type": "xsd:dateTime",
            },
            "issuer": Object {
              "@id": "cred:issuer",
              "@type": "@id",
            },
            "nonce": "sec:nonce",
            "proof": Object {
              "@id": "sec:proof",
              "@type": "@id",
            },
            "schema": "http://schema.org/",
            "sec": "https://w3id.org/security#",
            "signatureValue": "sec:signatureValue",
            "type": "@type",
            "xsd": "http://www.w3.org/2001/XMLSchema#",
          },
          Object {
            "ProofOfTinkererCredential": "https://identity.jolocom.com/terms/ProofOfTinkererCredential",
          },
        ],
        "claim": Object {
          "id": "did:jolo:5449f86ba2d4016169350a9766363d63a5326c0d2e8b922c2a75a803dd35d801",
          "note": "Thank you for attending our session at Web3!",
        },
        "expires": "2019-10-24T16:51:49.972Z",
        "id": "claimId:ac975ea8ccdf7",
        "issued": "2018-10-24T16:51:49.972Z",
        "issuer": "did:jolo:792b106ce93bc9e200f7812d33dbc3a367c57a0d5af11d0935eae1ca13c5c167",
        "name": "Tinkerer",
        "proof": Object {
          "created": "2018-10-24T16:51:49.972Z",
          "creator": "did:jolo:792b106ce93bc9e200f7812d33dbc3a367c57a0d5af11d0935eae1ca13c5c167#keys-1",
          "nonce": "93fab1092d1e7",
          "signatureValue": "kiowuS7ye6KFpeydfQM3oVicT2toiz65lb0DtR+gKoRhC4QLNCuBbcNR9zHJF9c3V2GeJ+/HqgzgQkMO+WOeQQ==",
          "type": "EcdsaKoblitzSignature2016",
        },
        "type": Array [
          "Credential",
          "ProofOfTinkererCredential",
        ],
      },
    ],
    "type": "SET_EXTERNAL",
  },
  Object {
    "loading": false,
    "type": "SET_LOADING",
  },
  Object {
    "routeName": "CredentialDialog",
    "type": "Navigation/NAVIGATE",
  },
]
`;

exports[`SSO action creators should correctly cancel submission of claims 1`] = `
Array [
  Object {
    "type": "CLEAR_CREDENTIAL_REQUEST",
  },
  Object {
    "actions": Array [
      Object {
        "routeName": "Home",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "key": undefined,
    "type": "Navigation/RESET",
  },
]
`;

exports[`SSO action creators should correctly cancel/go back from an SSO action  1`] = `
Array [
  Object {
    "type": "RESET_SELECTED",
  },
  Object {
    "actions": Array [
      Object {
        "routeName": "Home",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "key": undefined,
    "type": "Navigation/RESET",
  },
]
`;

exports[`SSO action creators should correctly consume Credential Requests 1`] = `
Array [
  Object {
    "type": "SET_CREDENTIAL_REQUEST",
    "value": Object {
      "availableCredentials": Array [
        Object {
          "type": "Name",
          "values": Array [
            "test",
            "test",
          ],
          "verifications": Array [
            Object {
              "expires": 2019-10-15T19:59:38.406Z,
              "id": "claimId:61adc6e7b1448",
              "issuer": "did:jolo:b310d293aeac8a5ca680232b96901fe85988fde2860a1a5db69b49762923cc88",
              "selfSigned": false,
            },
          ],
        },
      ],
      "callbackURL": "http://192.168.43.98:9000/authentication/3ktcr",
      "requester": "did:jolo:b310d293aeac8a5ca680232b96901fe85988fde2860a1a5db69b49762923cc88#keys-1",
    },
  },
  Object {
    "loading": false,
    "type": "SET_LOADING",
  },
  Object {
    "routeName": "Consent",
    "type": "Navigation/NAVIGATE",
  },
]
`;
